File: src/main/java/org/jetbrains/pmdkotlin/lang/kotlin/KotlinParser.java
package org.jetbrains.pmdkotlin.lang.kotlin;

import com.intellij.lang.ASTNode;
import com.intellij.lang.PsiBuilder;
import com.intellij.lang.PsiParser;
import com.intellij.lang.impl.PsiBuilderImpl;
import com.intellij.psi.PsiElement;
import com.intellij.psi.impl.source.tree.FileElement;
import net.sourceforge.pmd.lang.Parser;
import net.sourceforge.pmd.lang.ParserOptions;
import net.sourceforge.pmd.lang.TokenManager;
import net.sourceforge.pmd.lang.ast.ParseException;
import org.jetbrains.kotlin.KtNodeTypes;
import org.jetbrains.pmdkotlin.lang.kotlin.ast.AbstractKotlinNode;
import org.jetbrains.pmdkotlin.lang.kotlin.ast.KotlinASTNodeAdapter;

import java.io.Reader;
import java.util.HashMap;
import java.util.Map;

public class KotlinParser implements Parser {
    private org.jetbrains.kotlin.parsing.KotlinParser parser;

    private final ParserOptions parserOptions;

    public KotlinParser(ParserOptions parserOptions) {
        this.parserOptions = parserOptions;
        PsiParser intendedParser = KotlinFileContext.parserDefinition.createParser(KotlinFileContext.project);
        if (intendedParser instanceof org.jetbrains.kotlin.parsing.KotlinParser) {
            parser = (org.jetbrains.kotlin.parsing.KotlinParser) intendedParser;
        } else {
            throw new IllegalStateException("KotlinParser not found!");
        }
    }


    @Override
    public ParserOptions getParserOptions() {
        return this.parserOptions;
    }

    @Override
    public TokenManager getTokenManager(String s, Reader reader) {
        return new KotlinTokenManager(new KotlinFileContext(s, reader));
    }

    @Override
    public boolean canParse() {
        return true;
    }

    private void showTree(ASTNode v, String s) {
        //System.err.println(s + v.getClass());
        if (!(v instanceof FileElement)) {
            PsiElement psi = v.getPsi();

            System.err.println(s + v.getPsi().getClass());
        }
        for (ASTNode cv : v.getChildren(null)) {
            showTree(cv, s + "  ");
        }
    }

    @Override
    public AbstractKotlinNode parse(String fileName, Reader source) throws ParseException {
        KotlinFileContext kotlinContext = new KotlinFileContext(fileName, source);
        try {
            PsiBuilder builder = new PsiBuilderImpl(KotlinFileContext.project, kotlinContext.psiFile, KotlinFileContext.parserDefinition, new KotlinTokenManager(), null, kotlinContext.sourceCodeToString(), null, null);
            FileElement root = (FileElement) parser.parse(KtNodeTypes.KT_FILE, builder, kotlinContext.psiFile);
            root.setPsi(kotlinContext.psiFile.getNode().getPsi());

//            showTree(root, "Parsed tree");

            return new KotlinASTNodeAdapter(root.getPsi(), kotlinContext);
        } catch (Exception e) {
            throw new RuntimeException(e);
        }
    }

    @Override
    public Map<Integer, String> getSuppressMap() {
        return new HashMap<Integer, String>();
    }
}

File: src/main/kotlin/org/jetbrains/pmdkotlin/lang/kotlin/KotlinParser.kt
package org.jetbrains.pmdkotlin.lang.kotlin

import com.intellij.lang.ASTNode
import com.intellij.lang.impl.PsiBuilderImpl
import com.intellij.psi.impl.source.tree.FileElement
import net.sourceforge.pmd.lang.Parser
import net.sourceforge.pmd.lang.ParserOptions
import net.sourceforge.pmd.lang.TokenManager
import net.sourceforge.pmd.lang.ast.ParseException
import org.jetbrains.kotlin.KtNodeTypes
import org.jetbrains.kotlin.psi.KtElement
import org.jetbrains.pmdkotlin.lang.kotlin.ast.AbstractKotlinNode
import org.jetbrains.pmdkotlin.lang.kotlin.ast.KotlinASTNodeAdapter
import java.io.Reader

class KotlinParser(private val parserOptions: ParserOptions) : Parser {
    private var parser: org.jetbrains.kotlin.parsing.KotlinParser? = null

    init {
        val intendedParser = KotlinFileContext.parserDefinition.createParser(KotlinFileContext.project)
        if (intendedParser is org.jetbrains.kotlin.parsing.KotlinParser) {
            parser = intendedParser
        } else {
            throw IllegalStateException("KotlinParser not found!")
        }
    }

    override fun getParserOptions(): ParserOptions {
        return this.parserOptions
    }

    override fun getTokenManager(s: String, reader: Reader): TokenManager {
        return KotlinTokenManager(KotlinFileContext(s, reader))
    }

    override fun canParse(): Boolean {
        return true
    }

    private fun showTree(v: ASTNode, prefix: String) {
        if (v !is FileElement) {

            System.out.println("$prefix${v.psi.javaClass.simpleName} (${v.psi.text.replace("\n", "\\n").replace("\t", " ").replace("  ", " ")}) [${v.psi is KtElement}]")
        }
        for (cv in v.getChildren(null)) {
            showTree(cv, prefix + "  ")
        }
    }

    @Throws(ParseException::class)
    override fun parse(fileName: String, source: Reader): AbstractKotlinNode {
        val kotlinContext = KotlinFileContext(fileName, source)
        try {
            val builder = PsiBuilderImpl(KotlinFileContext.project, kotlinContext.psiFile, KotlinFileContext.parserDefinition, KotlinTokenManager(kotlinContext), null, kotlinContext.getSourceCode(), null, null)
            val root = parser?.parse(KtNodeTypes.KT_FILE, builder, kotlinContext.psiFile) as FileElement
            root.psi = kotlinContext.psiFile.node.psi

            showTree(root, "|")

            return KotlinASTNodeAdapter(root.psi, kotlinContext)
        } catch (e: Exception) {
            throw RuntimeException(e)
        }
    }

    override fun getSuppressMap(): Map<Int, String> {
        return HashMap()
    }
}
